cp_diffdock_service.py

```diff
@@ -0,0 +1,188 @@
+import asyncio
+import json
+import logging
+import re
+import os
+from websockets.server import serve
+import concurrent.futures
+from argparse import ArgumentParser
+
+from cp_diffdock_protocol import DiffDockProtocol
+from cp_diffdock_api import DiffDockApi
+from cp_ws_helpers import extractWsAppMessage, formWsAppMessage, formCompletedMessage
+
+logging.basicConfig(level=logging.INFO)
+log = logging.getLogger('diffdock_service')
+
+#
+# This is a websocket server for servicing DiffDock requests.
+# This listens on a port and places incoming json requests into a queue, serviced by a number of workers.
+#
+
+async def handleRequest(websocket, queue):
+    """
+    Receive a request and put it into the queue.
+    Make sure we keep the websocket so we can send the response when docking finishes.
+    """
+    try:
+        message = await websocket.recv()
+    except Exception as ex:
+        log.error(f"Exception reading from websocket: {ex}. Can't handle this request.")
+        return
+
+    log.info("Received a request.")
+    # Optional special handling for messages from bfd-server or other Bioleap WsApps
+    # requestId is non-empty if it is from a wsApp, and will need to be included in responses.
+    requestId, cmdName, requestData = extractWsAppMessage(message)
+    requestContext = websocket, requestId
+
+    # Parse
+    try:
+        requestObj = DiffDockProtocol.Request.from_json(requestData)
+    except:
+        log.exception(f"Rejected a request:\nMessage: {message}\nrequestData: {requestData}")
+        await sendError(requestContext, f"Invalid request")
+        return
+
+    # Enqueue
+    log.info("handleRequest running a request...")
+    try:
+        # Put the request into the queue with a condition to resolve when it finishes.
+        # Condition protects the websocket from closing early.
+        condition = asyncio.Condition()
+        await queue.put((requestObj, requestContext, condition))
+        await sendStatus(requestContext, "Request received")
+        async with condition:
+            await condition.wait()
+    except Exception as ex:
+        log.exception(f"handleRequest encountered exception")
+        await sendError(requestContext, f"Service exception: {ex}")
+
+
+async def queueWorker(queue):
+    """
+    Handle docking requests from the queue
+    """
+    with concurrent.futures.ThreadPoolExecutor() as executor:
```

Comment:
Does each worker get its own pool of threads? The doc about "reuses idle worker threads before starting _ worker threads" suggests that the workers all share, but are worker threads fungible?

Reply:
Hmm...yeah, this is a good catch.

Reply:
I'm going to hold off on making any changes for this. Moving `with ... as executor:` to main requires keeping the executor around for the lifetime of the workers and I don't really want to get into that right now.